File: ts/src/day2/day2_b.ts
import * as fs from "fs";
import readline from "readline";

async function mulPosDepth(): Promise<number> {
	const rl = readline.createInterface(fs.createReadStream("./input"));
	let pos = 0;
	let depth = 0;
	let aim = 0;

	for await (const line of rl) {
		const instruction = line.split(" ");

		const direction = instruction[0];
		const unit = Number.parseInt(instruction[1]);

		if (direction == "forward") {
			pos += unit;
			depth += aim * unit;
		} else if (direction == "down") {
			aim += unit;
		} else if (direction == "up") {
			aim -= unit;
		}
	}

	return new Promise<number>((resolve) => resolve(pos * depth));
}

(function main() {
	mulPosDepth().then(console.log);
})();

File: golang/day2/day2_b.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
	"strings"
)

func mulPosDepth() int {
	pos := 0
	depth := 0
	aim := 0

	file, _ := os.Open("./input")
	scanner := bufio.NewScanner(file)

	for scanner.Scan() {
		instruction := strings.Split(scanner.Text(), " ")
		unit, _ := strconv.Atoi(instruction[1])

		switch direction := instruction[0]; direction {
		case "forward":
			pos += unit
			depth += aim * unit
		case "down":
			aim += unit
		case "up":
			aim -= unit
		}
	}

	return pos * depth
}

func main() {
	fmt.Println(mulPosDepth())
}
